contrib/masterfiles-stage/common.sh

```diff
@@ -12,175 +12,252 @@ if [ ! "$debug_mode" == "true" ]; then
-  fi
+  # Depends on $local_mirrored_repo
+  # Accepts two args:
+  # $1 - dir to deploy to
+  # $2 - refspec to deploy - a git tagname, branch, or commit hash.
 
-  if /var/cfengine/bin/cf-promises -T "${STAGING_DIR}"; then
-    if ! /var/cfengine/bin/cf-promises -cf "${STAGING_DIR}/update.cf"; then
-        error_exit "Update policy staged in ${STAGING_DIR} could not be validated, aborting."
-    fi
-      # roll out the release if the release IDs are different
-#      if /usr/bin/diff -q "${STAGING_DIR}/cf_promises_release_id" "${MASTERDIR}/cf_promises_release_id" ; then
-#          #echo "No release needs to be made, the release IDs are the same."
-#          touch "${STAGING_DIR}"
-#      else
-          cd "${STAGING_DIR}" && (
-          chown -R root:root "${STAGING_DIR}" && rsync -rltDE -c --delete-after --chmod=u+rwX,go-rwx "${STAGING_DIR}/" "${MASTERDIR}/" && echo "Successfully deployed commit '${GIT_TAG_OR_COMMIT}' from '${GIT_URL}' to '${MASTERDIR}' on $(date)"
-      )
-#      fi
+  # This function
+  # 1. creates an empty temp dir,
+  # 2. checks out the refspec into the empty temp dir,
+  #    (including populating .git/HEAD in the temp dir),
+  # 3. sets appropriate permissions on the policy set,
+  # 4. validates the policy set using cf-promises,
+  # 5. moves the temp dir policy set into the given deploy dir,
+  #    avoiding triggering policy updates unnecessarily
+  #    by comparing the cf_promises_validated flag file.
+  #    (See long comment at end of function def.)
+
+  # Ensure absolute pathname is given
+  [ "${1:0:1}" = / ] ||
+    error_exit "You must specify absolute pathnames in channel_config: '$1'"
+  mkdir -p "$(dirname "$1")" || error_exit "Failed to mkdir -p dirname $1"
+    # We don't mkdir $1 directly, just its parent dir if that doesn't exist.
+
+  ########################## 1. CREATE EMPTY TEMP DIR
+  # Put staging dir right next to deploy dir to ensure it's on same filesystem
+  local temp_stage
+  temp_stage="$(mktemp -d --tmpdir="$(dirname "$1")" )"
+
+  ########################## 2. CHECKOUT INTO TEMP DIR
+  # The '^0' at the end of the refspec
+  # populates HEAD with the SHA of the commit
+  # rather than potentially using a git branch name.
+  # Also see http://stackoverflow.com/a/13293010/5419599
+  # and https://github.com/cfengine/core/pull/2465#issuecomment-173656475
+  git --git-dir="${local_mirrored_repo}" --work-tree="${temp_stage}" checkout -q -f "${2}^0" ||
+    error_exit "Failed to checkout '$2' from '${local_mirrored_repo}'"
+  # Grab HEAD so it can be used to populate cf_promises_release_id
+  mkdir -p "${temp_stage}/.git"
+  cp "${local_mirrored_repo}/HEAD" "${temp_stage}/.git/"
+
+  ########################## 3. SET PERMISSIONS ON POLICY SET
+  chown -R root:root "${temp_stage}" || error_exit "Unable to chown '${temp_stage}'"
+  chmod -R go-rwx    "${temp_stage}" || error_exit "Unable to chmod '${temp_stage}'"
+
```

Comment:
Not better than using `find`, but better than using `xargs`, which will [cause issues when filenames contain whitespace](http://unix.stackexchange.com/q/131766/135943).  Instead I'll fix this to do the whole action with `find` (using the `-exec` flag).  Thanks for pointing out the permissions issue; I agree it will be good to get rid of those warning messages.